### PR TITLE
RYA-85 Configuration glue code

### PR DESCRIPTION
**Pull request by isper3at**
## Description

Added a configuration utility for updating the
configuration used in Rya based on what is
in RyaDetails, the admin table.

Added integration examples, these will need to
be updated to use JUnit
##### Test Coverage

Is not complete since the current design makes this very hard.  Both accumulo and mongo have some test coverage though.
### Links

[Jira](https://issues.apache.org/jira/browse/RYA-85)
### Checklist
- [ ] Code Review
- [ ] Squash commits
#### People To Review

@kchilton2 
@pujav65
@meiercaleb

### CODE REVIEW

File: common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package mvm.rya.api.instance;
```
Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java:1]: I don't think global configuration values should be in this package.

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * add them to the application's {@link Configuration}.
+ */
+public class RyaDetailsToConfiguration {
+    private static final Logger LOG = Logger.getLogger(RyaDetailsToConfiguration.class);
+    public static final String CONF_PCJ_IDS = "pcj.index.ids";
+    /**
+     * @param details - The {@link RyaDetails} to add to the {@link Configuration}.
+     * @param conf - The {@link Configuration} to add {@link RyaDetails} to.
+     */
+    public static void addRyaDetailsToConfiguration(final RyaDetails details, final Configuration conf) {
```
Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java:23]: Maybe call this overwrite instead of add.

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * Used to fetch {@link RyaDetails} from a {@link RyaDetailsRepository} and
+ * add them to the application's {@link Configuration}.
+ */
+public class RyaDetailsToConfiguration {
+    private static final Logger LOG = Logger.getLogger(RyaDetailsToConfiguration.class);
+    public static final String CONF_PCJ_IDS = "pcj.index.ids";
+    /**
+     * @param details - The {@link RyaDetails} to add to the {@link Configuration}.
```
Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java:20]: Or document up here that any values that conflict will be overwritten.

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        checkAndSet(conf, ConfigurationFields.USE_TEMPORAL, details.getTemporalIndexDetails().isEnabled());
+        final boolean pcjEnabled = details.getPCJIndexDetails().isEnabled();
+        checkAndSet(conf, ConfigurationFields.USE_PCJ, pcjEnabled);
+        if(pcjEnabled) {
+            final PCJIndexDetails pcjDetails = details.getPCJIndexDetails();
+            final List<String> pcjIds = new ArrayList<>();
+            for(final PCJDetails pcj : pcjDetails.getPCJDetails()) {
+                pcjIds.add(pcj.getId());
+            }
+            conf.setStrings(CONF_PCJ_IDS, pcjIds.toArray(new String[]{}));
```
Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java]: Do these values get used anywhere? I don't see any code updates that would indicate just the IDs are required instead of the entire table names.

File: dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java [hunk 5]
```diff
@@ -276,7 +275,6 @@ public void update() throws AlreadyInitializedException, RyaDetailsRepositoryExc
         assertEquals(updated, fetched);
     }
 
-    @Test(expected = ConcurrentUpdateException.class)
```
Comment (kchilton2) [inline dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java:279]: Why did you remove the test annotation from this test?

File: extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import mvm.rya.api.instance.RyaDetails.PCJIndexDetails.PCJDetails.PCJUpdateStrategy;
+import mvm.rya.api.instance.RyaDetails.ProspectorDetails;
+import mvm.rya.api.instance.RyaDetails.TemporalIndexDetails;
+import mvm.rya.api.instance.RyaDetailsToConfiguration;
+import mvm.rya.api.persist.RyaDAOException;
+import mvm.rya.indexing.accumulo.ConfigUtils;
+import mvm.rya.indexing.accumulo.geo.GeoConstants;
+import mvm.rya.sail.config.RyaSailFactory;
+
+public class AccumuloRyaDetailsExample {
```
Comment (kchilton2) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: What exactly is the purpose of this example? Why not just make this an integration test?

File: extras/indexingExample/src/main/java/MongoRyaDetailsExample.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+import mvm.rya.api.instance.RyaDetailsToConfiguration;
+import mvm.rya.api.persist.RyaDAOException;
+import mvm.rya.indexing.accumulo.ConfigUtils;
+import mvm.rya.indexing.accumulo.geo.GeoConstants;
+import mvm.rya.mongodb.MongoDBRdfConfiguration;
+import mvm.rya.mongodb.MongoDBRyaDAO;
+import mvm.rya.rdftriplestore.RdfCloudTripleStore;
+import mvm.rya.sail.config.RyaSailFactory;
+
+public class MongoRyaDetailsExample {
```
Comment (kchilton2) [inline extras/indexingExample/src/main/java/MongoRyaDetailsExample.java]: Same question here. I think the configuration augmentation should be transparent to any user of our application so that however the example used to work shouldn't change. Or like, once you pass the configuration into our logic we do the changes if the details are present.

File: common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java (hunk 1 above)
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java:1]: i think i package scoped it,...if not, thoughts?

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java (hunk 2 above)
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java:23]: done

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java (hunk 3 above)
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java:20]: like this more

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java (hunk 4 above)
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java]: this was what I thought we were going to do.  I don't exactly know where these would get used

File: dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java (hunk 5 above)
Comment (isper3at) [inline dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java:279]: on the mongo side this exception is never thrown

File: extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java (hunk 6 above)
Comment (isper3at) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: I had intented to make it so.  This project doesn't have JUnity as a dep.

File: extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                        .build())
+            .setProspectorDetails( new ProspectorDetails(Optional.of(new Date())) )
+            .setJoinSelectivityDetails( new JoinSelectivityDetails(Optional.of(new Date())) );
+        RyaDetailsToConfiguration.addRyaDetailsToConfiguration(builder.build(), conf);
+
+        SailRepository repository = null;
+        SailRepositoryConnection conn = null;
+        try {
+            log.info("Connecting to Indexing Sail Repository.");
+            final Sail sail = RyaSailFactory.getInstance(conf);
```
Comment (meiercaleb) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: RyaSailFactory should use your RyaDetailsToConfiguration object to merge the user's configuration with the RyaDetails for the specified Rya prefix.  To get the RyaDetails for the instance, the RyaSailFactory should first determine whether the user wants to use Mongo or Accumulo.  Based on that configuration, the Factory should initialize the appropriate RyaDetailsRepository to fetch the details for merging the user's config with the prefix config.  Again, this should all happen within the RyaSailFactory and be hidden from the user.

File: extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java (hunk 6 above)
Comment (meiercaleb) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: This is more of a unit test for RyaDetailsToConfiguration than an example of how to configure a query client.  I would include it with your unit tests.  I don't think you need to create a new example.  You just need to integrate your changes along with Kevin's RyaDetailsRepository with the RyaSailFactory and make sure that the current examples work.  See my comment below.

Comment (isper3at) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: its an integration test, its not really supposed to be an example, this is just where I was told to put them.

Comment (kchilton2) [inline extras/indexingExample/src/main/java/AccumuloRyaDetailsExample.java]: I really don't think this is what Puja intended.

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 9]
```diff
@@ -30,57 +39,97 @@
...
+import mvm.rya.mongodb.instance.MongoRyaInstanceDetailsRepository;
 import mvm.rya.rdftriplestore.RdfCloudTripleStore;
 import mvm.rya.rdftriplestore.inference.InferenceEngine;
 import mvm.rya.rdftriplestore.inference.InferenceEngineException;
 
 public class RyaSailFactory {
-
-
-
+    @Deprecated
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: not deprecated

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 10]
```diff
@@ -5,9 +5,18 @@
...
+import org.apache.accumulo.core.client.security.tokens.PasswordToken;
 import org.apache.hadoop.conf.Configuration;
 import org.openrdf.sail.Sail;
 
+import com.google.common.collect.Lists;
+import com.mongodb.MongoClient;
+import com.mongodb.MongoCredential;
+import com.mongodb.ServerAddress;
+
+import jline.internal.Log;
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: log4j

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 11]
```diff
@@ -30,57 +39,97 @@
...
-            conf.setTablePrefix(config.get(RdfCloudTripleStoreConfiguration.CONF_TBL_PREFIX)); // sets
-                                                                                               // TablePrefixLayoutStrategy
-            ConfigUtils.setIndexers(conf);
-            conf.setDisplayQueryPlan(true);
-
-            crdfdao.setConf(conf);
-            crdfdao.init();
-            store.setRyaDAO(crdfdao);
+            try {
+                final PasswordToken pswdToken = new PasswordToken(pswd);
```
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: use if isInitialized()

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java [hunk 12]
```diff
@@ -39,6 +39,9 @@
     private static final Logger LOG = Logger.getLogger(RyaDetailsToConfiguration.class);
     public static final String CONF_PCJ_IDS = "pcj.index.ids";
     /**
+     * Ensures the values in the {@link Configuration} do not conflict with the values in {@link RyaDetails}.
+     * If they do, the values in {@link RyaDetails} take precedent and the {@link Configuration} value will
+     * be overwritten.
```
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java:18]: nit: newline

File: common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java [hunk 13]
```diff
@@ -22,7 +22,7 @@
 /**
  * The possible configuration fields used in a Rya application.
  */
-public class ConfigurationFields {
+class ConfigurationFields {
```
Comment (isper3at) [inline common/rya.api/src/main/java/mvm/rya/api/instance/ConfigurationFields.java:29]: todo: refactor, indicate where they came from

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 14]
```diff
@@ -36,12 +19,26 @@
...
 
+import java.net.UnknownHostException;
+
+import org.apache.accumulo.core.client.AccumuloException;
+import org.apache.accumulo.core.client.AccumuloSecurityException;
+import org.apache.accumulo.core.client.Connector;
+import org.apache.accumulo.core.client.Instance;
+import org.apache.accumulo.core.client.security.tokens.PasswordToken;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.log4j.Logger;
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: Use slf4j.

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 15]
```diff
@@ -55,91 +52,98 @@
...
-                    } catch (NumberFormatException | UnknownHostException e) {
-                        e.printStackTrace();
-                    }
-                }
-                crdfdao.init();
-
-                conf.setDisplayQueryPlan(true);
-                store.setRyaDAO(crdfdao);
-            }
+            final MongoClient client = updateMongoConfig((MongoDBRdfConfiguration) conf, user, pswd, instance);
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: Why do we pass the Accumulo Instance Name into this function?

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 16]
```diff
@@ -55,91 +52,98 @@
...
         return getRyaSail(conf);
     }
 
-    private static Sail getRyaSail(final Configuration config) throws AccumuloException, AccumuloSecurityException, RyaDAOException, InferenceEngineException {
+    private static Sail getRyaSail(final Configuration config) throws InferenceEngineException, RyaDAOException, AccumuloException, AccumuloSecurityException {
         final RdfCloudTripleStore store = new RdfCloudTripleStore();
-        RyaDAO crdfdao = null;
-        RdfCloudTripleStoreConfiguration conf;
+        final RyaDAO dao;
+        final RdfCloudTripleStoreConfiguration conf;
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: Having a variable named 'config' and one named 'conf' is really confusing. Is 'conf' an updated version of 'config'? How do they differ?

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 17]
```diff
@@ -55,91 +52,98 @@
...
-                conf.setTablePrefix(config.get(RdfCloudTripleStoreConfiguration.CONF_TBL_PREFIX)); // sets
-                // TablePrefixLayoutStrategy
-                ConfigUtils.setIndexers(conf);
-                conf.setDisplayQueryPlan(true);
-
-                crdfdao.setConf(conf);
-                crdfdao.init();
-                store.setRyaDAO(crdfdao);
-            }
+            updateAccumuloConfig((AccumuloRdfConfiguration) conf, user, pswd, instance);
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: You're using the wrong value here too. AccumuloRyaInstanceDetailsRepository requires the Rya instance name, not the Accumulo instance name. I think you're confused about what the 'ConfigUtils.CLOUDBASE_INSTANCE' field indicates.

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 18]
```diff
@@ -55,91 +52,98 @@
...
+            dao = getAccumuloDAO((AccumuloRdfConfiguration)conf);
         }
+        conf.setTablePrefix(config.get(RdfCloudTripleStoreConfiguration.CONF_TBL_PREFIX));
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: It looks like conf is a copy of config, so why do you need to do this explicitly?

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 15 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: is the Accumulo instance name different from the Rya Instance name?

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 16 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: one is the Hadoop config the other is the CloudTripleStoreConfiguration

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 17 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: entirely likely

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 14 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: done

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 18 above)
Comment (isper3at) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: That line was there before, so I just left it.

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java (hunk 15 above)
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java]: Yes. The Accumulo Instance Name is the the name that IDs Accumulo within Zookeeper. The Rya Instance Name is the name of the Rya instance hosted by Accumulo.

File: dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java [hunk 19]
```diff
@@ -276,7 +275,6 @@ public void update() throws AlreadyInitializedException, RyaDetailsRepositoryExc
...
     public void update_outOfDate() throws AlreadyInitializedException, RyaDetailsRepositoryException {
```
Comment (kchilton2) [inline dao/mongodb.rya/src/test/java/mvm/rya/mongodb/instance/MongoRyaDetailsRepositoryIT.java:280]: Fix the test.

File: common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+import mvm.rya.api.instance.RyaDetails.PCJIndexDetails;
+import mvm.rya.api.instance.RyaDetails.PCJIndexDetails.PCJDetails;
+
+/**
+ * Used to fetch {@link RyaDetails} from a {@link RyaDetailsRepository} and
+ * add them to the application's {@link Configuration}.
+ */
+public class RyaDetailsToConfiguration {
+    private static final Logger LOG = Logger.getLogger(RyaDetailsToConfiguration.class);
+    public static final String CONF_PCJ_IDS = "pcj.index.ids";
```
Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java]: This should probably be in ConfigurationFields.

Comment (kchilton2) [inline common/rya.api/src/main/java/mvm/rya/api/instance/RyaDetailsToConfiguration.java]: Just kidding, this isn't what I thought it was. I think we can delete this entirely for this review.

File: extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java [hunk 21]
```diff
@@ -27,73 +19,132 @@
...
-            crdfdao = new MongoDBRyaDAO((MongoDBRdfConfiguration)conf);
-            crdfdao.init();
-
-            conf.setDisplayQueryPlan(true);
-            store.setRyaDAO(crdfdao);
+        final RyaDAO dao;
+        final RdfCloudTripleStoreConfiguration rdfConfig;
+
+        final String user = config.get(ConfigUtils.CLOUDBASE_USER);
+        final String pswd = config.get(ConfigUtils.CLOUDBASE_PASSWORD);
```
Comment (kchilton2) [inline extras/indexing/src/main/java/mvm/rya/sail/config/RyaSailFactory.java:69]: Could we make sure the "CLOUDBASE" username and password are also meant to be used for Mongo? That isn't very intuitive.
